flink-streaming-java/src/test/java/org/apache/flink/streaming/api/graph/StreamingJobGraphGeneratorTest.java

```diff
@@ -657,6 +660,117 @@ public void testNotSupportInputSelectableOperatorIfCheckpointing() {
+
+		final StreamConfig map1Config = Iterables.getOnlyElement(
+			sourceConfig.getTransitiveChainedTaskConfigs(StreamingJobGraphGeneratorTest.class.getClassLoader()).values());
+		assertEquals(0.3, map1Config.getManagedMemoryFractionOnHeap(), 0.000001);
+		assertEquals(0.5, map1Config.getManagedMemoryFractionOffHeap(), 0.000001);
+
+		final StreamConfig map2Config = new StreamConfig(vertex2.getConfiguration());
+		assertEquals(0.6, map2Config.getManagedMemoryFractionOnHeap(), 0.000001);
+		assertEquals(0.1, map2Config.getManagedMemoryFractionOffHeap(), 0.000001);
+
+		final StreamConfig map3Config = new StreamConfig(vertex3.getConfiguration());
+		assertEquals(1.0, map3Config.getManagedMemoryFractionOnHeap(), 0.000001);
+		assertEquals(1.0, map3Config.getManagedMemoryFractionOffHeap(), 0.000001);
+	}
+
+	@Test
+	public void testManagedMemoryFractionForUnknownResourceSpec() throws Exception {
+		final ResourceSpec resource1 = ResourceSpec.UNKNOWN;
+		final ResourceSpec resource2 = ResourceSpec.UNKNOWN;
+		final ResourceSpec resource3 = ResourceSpec.UNKNOWN;
+		final ResourceSpec resource4 = ResourceSpec.UNKNOWN;
+
+		final JobGraph jobGraph = createJobGraphForManagedMemoryFractionTest(resource1, resource2, resource3, resource4);
+		final JobVertex vertex1 = jobGraph.getVerticesSortedTopologicallyFromSources().get(0);
+		final JobVertex vertex2 = jobGraph.getVerticesSortedTopologicallyFromSources().get(1);
+		final JobVertex vertex3 = jobGraph.getVerticesSortedTopologicallyFromSources().get(2);
+
+		final StreamConfig sourceConfig = new StreamConfig(vertex1.getConfiguration());
+		assertEquals(1.0 / 3, sourceConfig.getManagedMemoryFractionOnHeap(), 0.000001);
+		assertEquals(1.0 / 3, sourceConfig.getManagedMemoryFractionOffHeap(), 0.000001);
+
+		final StreamConfig map1Config = Iterables.getOnlyElement(
+			sourceConfig.getTransitiveChainedTaskConfigs(StreamingJobGraphGeneratorTest.class.getClassLoader()).values());
+		assertEquals(1.0 / 3, map1Config.getManagedMemoryFractionOnHeap(), 0.000001);
+		assertEquals(1.0 / 3, map1Config.getManagedMemoryFractionOffHeap(), 0.000001);
+
+		final StreamConfig map2Config = new StreamConfig(vertex2.getConfiguration());
+		assertEquals(1.0 / 3, map2Config.getManagedMemoryFractionOnHeap(), 0.000001);
+		assertEquals(1.0 / 3, map2Config.getManagedMemoryFractionOffHeap(), 0.000001);
+
+		final StreamConfig map3Config = new StreamConfig(vertex3.getConfiguration());
+		assertEquals(1.0, map3Config.getManagedMemoryFractionOnHeap(), 0.000001);
+		assertEquals(1.0, map3Config.getManagedMemoryFractionOffHeap(), 0.000001);
+
+	}
+
+	private JobGraph createJobGraphForManagedMemoryFractionTest(
+		final ResourceSpec resource1,
+		final ResourceSpec resource2,
+		final ResourceSpec resource3,
+		final ResourceSpec resource4) throws Exception {
+
+		final Method opMethod = SingleOutputStreamOperator.class.getDeclaredMethod("setResources", ResourceSpec.class);
```

Comment:
let's put this into a re-usable method; I found already multiple instances of the same access pattern and we shouldn't spread around reflection calls.

Reply:
done.